### PR TITLE
Awesome datastore

### LINKED ISSUE #229: daemon leaves behind api file lock on error
When the daemon hits an error on load like https://github.com/ipfs/js-ipfs/issues/228, it will leave around the `~/.ipfs/api` file lock.

`go-ipfs` seems to be smart about detecting this and cleaning it up, but it will inhibit future invocations of the `js-ipfs` daemon.

### PR DESCRIPTION
**Pull request by dignifiedquire**
Closes https://github.com/ipfs/js-ipfs/issues/787
Closes #229

- [x] refactored cli to use a context ipfs daemon to ensure cleanup
- [x] update to new datastore dependencies
- [x] fix timeout issues and add two new events (`init` and `ready`) to handle the startup of a daemon more granularly (needs discussion probably)
- [x] upgrade to yargs@7

### CODE REVIEW

**Comment by dignifiedquire** [comment]
I am afraid this is not ready yet. I've discovered that the new creation and startup api is the source of quite some race conditions, which are becoming more visible with this PR, as the locking of the repo and open and close are more enforced. 

We need to think through this very carefully as the options `init` and `start` combined with methods for doing those actions are nice for the user but create quite the tricky state management in the background. If we decide to keep the api we need to integrate a proper finite state machine to ensure all transitions are valid and very explicit descriptions on how things work. 

Tldr `setTimeout` usually means you have a race condition you are hiding..

File: src/cli/bin.js [hunk 1]
```diff
@@ -27,9 +28,23 @@ aliases.forEach((alias) => {
   cli.command(alias.command, alias.describe, alias.builder, alias.handler)
 })
 
-// finalize cli setup
-cli // eslint-disable-line
-  .help()
-  .strict()
-  .completion()
-  .argv
+utils.getIPFS((err, ipfs, cleanup) => {
```
Comment (diasdavid) [inline src/cli/bin.js]: init and daemon don't use an `ipfs` instance to run, how does this cover those two cases?

**Comment by diasdavid** [comment]
Note: We can add a `init` event to the set up, it should solve all of the issues you are seeing as the only issue was when a repo is init'ed by default and gets started 'later', this later is the source of trouble because there is no way to monitor when the init has finished.

The finite state machine approach is also interesting, probably even better to keep the simplicity. Wanna take a stab at it?

**Comment by diasdavid** [comment]
See: https://github.com/ipfs/js-ipfs/compare/fix/race?expand=1

File: src/cli/bin.js (hunk 1 above)
Comment (dignifiedquire) [inline src/cli/bin.js]: I will add an exception to those, I didn't realise that until later

**Comment by diasdavid** [comment]
Saw now that you added more events.

- `init` is part of master now
- `ready` why the need for this one, exactly?

**Comment by dignifiedquire** [comment]
> ready why the need for this one, exactly?

to detect the finishing things  like `new IPFS({init: false, start: false}` which are alway async now, due to the need to detect and open the repo

**Comment by diasdavid** [comment]
@dignifiedquire in that case, there are no actions happening asynchronously.

**Comment by dignifiedquire** [comment]
> @dignifiedquire in that case, there are no actions happening asynchronously.

but they are now, because repo checks are async

File: gulpfile.js [hunk 2]
```diff
@@ -16,37 +15,25 @@ let nodes = []
...
+        `/ip4/127.0.0.1/tcp/10${num}`,
+        `/ip4/127.0.0.1/tcp/20${num}/ws`
+      ],
+      API: `/ip4/127.0.0.1/tcp/31${num}`,
+      Gateway: `/ip4/127.0.0.1/tcp/32${num}`
     },
-    start: false,
-    EXPERIMENTAL: {
-      pubsub: true
-    },
```
Comment (diasdavid) [inline gulpfile.js]: why disabling PubSub?

File: src/cli/bin.js [hunk 3]
```diff
@@ -27,9 +28,30 @@ aliases.forEach((alias) => {
...
-  .help()
-  .strict()
-  .completion()
-  .argv
+const args = process.argv.slice(2)
+
+// Need to skip to avoid locking the daemon
+if (args[0] === 'daemon' || args[0] === 'init') {
+  return cli.help().strict(false).completion().parse(args)
+}
```
Comment (diasdavid) [inline src/cli/bin.js]: Make this more clear:

`// commands that don't require a daemon/instance`
`// commands that require a daemon/instance`

Comment (diasdavid) [inline src/cli/bin.js]: Also, use same indentation in both

File: src/core/components/block.js [hunk 4]
```diff
@@ -11,43 +12,32 @@ module.exports = function block (self) {
...
-
-        callback(null, block)
-      })
+        (block, cb) => self._blockService.put(block, (err) => {
+          if (err) {
+            return cb(err)
+          }
+          cb(null, block)
+        })
+      ], callback)
```
Comment (diasdavid) [inline src/core/components/block.js]: This has changed the API, block needs to be able to:

```
» jsipfs block put --help
jsipfs block put [block]

Options:
  --help        Show help                                              [boolean]
  --format, -f  cid format for blocks to be created with.    [default: "dag-pb"]
  --mhtype      multihash hash function                    [default: "sha2-256"]
  --mhlen       multihash hash length
```

Before we passed that through CID, if you remove the CID from equation and always calculate the multihash of sha256, then you can't block put eth data anymore.

File: src/core/components/libp2p.js [hunk 5]
```diff
@@ -14,9 +15,9 @@ module.exports = function libp2p (self) {
         }
 
         const options = {
-          mdns: config.Discovery.MDNS.Enabled,
-          webRTCStar: config.Discovery.webRTCStar.Enabled,
-          bootstrap: config.Bootstrap
+          mdns: get(config, 'Discovery.MDNS.Enabled'),
+          webRTCStar: get(config, 'Discovery.webRTCStar.Enabled'),
+          bootstrap: get(config, 'Bootstrap')
```
Comment (diasdavid) [inline src/core/components/libp2p.js]: Requiring a dependency to pick one property of an object?

File: src/core/components/pre-start.js [hunk 6]
```diff
@@ -31,7 +30,7 @@ module.exports = function preStart (self) {
 
           if (!mafmt.IPFS.matches(ma)) {
             ma = ma.encapsulate('/ipfs/' +
-              self._peerInfo.id.toB58String())
+                                self._peerInfo.id.toB58String())
```
Comment (diasdavid) [inline src/core/components/pre-start.js]: strange identation

File: src/core/status.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+// Possible IPFS status values:
+module.exports = {
+  new: 'new',
+  initializing: 'initializing',
+  initialized: 'initialized',
+  starting: 'starting',
+  running: 'running',
+  stopping: 'stopping',
+  stopped: 'stopped'
+}
```
Comment (diasdavid) [inline src/core/status.js]: Right now, in order to follow the sequence of booting a node, we need to open 4+ files. 

Let's put this into the state.js.

Comment (diasdavid) [inline src/core/status.js]: Rechecked, this is just used to populate a var `this._status` once and that const is not used anymore. ???

File: src/core/utils.js [hunk 8]
```diff
@@ -1,17 +1,3 @@
...
-    }
-
-    if (exists === false) {
-      return cb(new Error('no ipfs repo found'))
-    }
-
-    cb()
-  })
-}
-
```
Comment (diasdavid) [inline src/core/utils.js]: Where is this now? Breaking API change? If so, update README

File: src/http-api/index.js [hunk 9]
```diff
@@ -22,56 +16,61 @@ function uriToMultiaddr (uri) {
...
+            init: init,
+            start: true,
+            config: config,
+            EXPERIMENTAL: {
+              pubsub: true
+            }
+          })
+        } catch (err) {
+          return cb(err)
+        }
```
Comment (diasdavid) [inline src/http-api/index.js]: What is throwing? Errors should come on the `error` event only.

File: test/go-ipfs-repo/datastore/LOG [hunk 10]
```diff
@@ -0,0 +1,2 @@
+=============== Mar 18, 2017 (CET) ===============
+00:21:30.447664 log@legend F·NumFile S·FileSize N·Entry C·BadEntry B·BadBlock Ke·KeyError D·DroppedEntry L·Level Q·SeqNum T·TimeElapsed
```
Comment (diasdavid) [inline test/go-ipfs-repo/datastore/LOG]: .gitignore this

File: test/go-ipfs-repo/version [hunk 11]
```diff
@@ -0,0 +1 @@
+1
```
Comment (diasdavid) [inline test/go-ipfs-repo/version]: is it really version 1?

File: gulpfile.js (hunk 2 above)
Comment (dignifiedquire) [inline gulpfile.js]: It's not disabled, it is enabled in the httpserver as far as I understand

File: src/core/components/libp2p.js (hunk 5 above)
Comment (dignifiedquire) [inline src/core/components/libp2p.js]: This is so that undefined chains don't throw. This code explodes if you try to read a config from go-ipfs as it doesn't have a `webRTCStar` field in discovery.

File: src/core/components/pre-start.js (hunk 6 above)
Comment (dignifiedquire) [inline src/core/components/pre-start.js]: very strange

File: src/core/status.js (hunk 7 above)
Comment (dignifiedquire) [inline src/core/status.js]: this is the rest from some experimentation, going to delete it

File: src/core/utils.js (hunk 8 above)
Comment (dignifiedquire) [inline src/core/utils.js]: I didn't know this was in the readme, this is not needed anymore internally that's why I removed it

File: src/http-api/index.js (hunk 9 above)
Comment (dignifiedquire) [inline src/http-api/index.js]: this can throw if someone screwed up in the implementation, resulting in silent fail of the tests

File: test/go-ipfs-repo/version (hunk 11 above)
Comment (dignifiedquire) [inline test/go-ipfs-repo/version]: hmm no not sure why that says version 1

File: src/core/utils.js (hunk 8 above)
Comment (dignifiedquire) [inline src/core/utils.js]: Can't find it in the readme, are you sure this was documented somewhere?

File: gulpfile.js (hunk 2 above)
Comment (diasdavid) [inline gulpfile.js]: got it, true.

File: src/core/components/block.js (hunk 4 above)
Comment (dignifiedquire) [inline src/core/components/block.js]: I don't understand your concern, if you want to specify your own things you pass a regular ipfs block  here which already defines the cid it wants, this way things are passed through as expected

Comment (dignifiedquire) [inline src/core/components/block.js]: the hashing is just a fallback if only a buffer is passed in.

File: src/core/utils.js (hunk 8 above)
Comment (diasdavid) [inline src/core/utils.js]: Ah, not on the readme, people are using it though - https://github.com/ipfs/js-ipfs/pull/776

Wanna finish that PR and merge it here?

File: test/go-ipfs-repo/version (hunk 11 above)
Comment (dignifiedquire) [inline test/go-ipfs-repo/version]: on disk it says 5 for me

Comment (dignifiedquire) [inline test/go-ipfs-repo/version]: fixed, there were two

File: test/go-ipfs-repo/datastore/LOG (hunk 10 above)
Comment (dignifiedquire) [inline test/go-ipfs-repo/datastore/LOG]: done

File: src/http-api/index.js (hunk 9 above)
Comment (diasdavid) [inline src/http-api/index.js]: ok, let's add that comment, I want to avoid giving the idea that errors should be 'thrown' for contributors coming into the codebase.

File: src/core/utils.js (hunk 8 above)
Comment (dignifiedquire) [inline src/core/utils.js]: lets get this merged and then I can do that PR, I don't want to put even more things into here than there are already

File: src/http-api/index.js (hunk 9 above)
Comment (dignifiedquire) [inline src/http-api/index.js]: done

File: src/core/components/block.js (hunk 4 above)
Comment (diasdavid) [inline src/core/components/block.js]: Hm.. I see, you now expect everything in the block.

Can we also have the options object, we will need to support the hashAlg and format. But that can be a separate PR

**Comment by diasdavid** [comment]
Seems that last thing is to rebase

**Comment by diasdavid** [comment]
image

I need to leave now for up to 3 hours, will come back online later.

@dignifiedquire, let me know if you can check this.

**Comment by dignifiedquire** [comment]
last missing piece https://github.com/ipfs/js-ipfs-repo/pull/128 (hopefully)
